### PR TITLE
DPE-2225 using allocable memory instead of `free`

### LINKED ISSUE #254: mysql-k8s pod gets recreated on `update-status` (under the heavy load): juju.worker.caasunitterminationworker terminating due to SIGTERM
## Steps to reproduce
1) Deploy mysql-k8s + mysql-router-k8s on GKE using https://discourse.charmhub.io/t/charmed-mysql-k8s-how-to-deploy-on-gke/10875
2) Run performance testing using https://discourse.charmhub.io/t/charmed-mysql-k8s-how-to-performance-test/11073

## Expected behavior
The performance test must finish successfully.

## Actual behavior
The test fails with an error:
```
FATAL: mysql_drv_query() returned error 2013 (Lost connection to MySQL server during query) for query 'INSERT INTO ..."
```
Literally MySQL K8s Primary has gone away. K8s says the pod has been recreated (not-even restarted):
```
> kubectl get pods -n mysql-k8s
NAME                             READY   STATUS    RESTARTS   AGE   IP 
mysql-k8s-0                      2/2     Running   0          11h   10.104.2.10
mysql-k8s-1                      2/2     Running   0          28m   10.104.1.12 <<<<<<<<<<<< 28m, all units deployed simultaneously!
mysql-k8s-2                      2/2     Running   0          11h   10.104.0.4

> kubectl describe pod mysql-k8s-1 -n mysql-k8s
...
Containers:
...
  mysql:
...
    State:          Running
      Started:      Mon, 10 Jul 2023 23:44:45 +0200
    Ready:          True
    Restart Count:  0  <<<<<<<<<<<<<<<<<<
```

Juju debug logs shows a try to run `update-status` on the MySQL Primary node.

## Versions

Operating system: 22.04
Juju CLI: 2.9.44
Juju agent: 2.9.44
Charm revision: 8.0/edge
microk8s: Tested on GKE and   microk8s  v1.27.2   5372   1.27/stable    canonical✓  classic

## Log output
Juju debug log: 
```
[...skipped non-mysql-k8s-0 log lines...]
unit-mysql-k8s-0: 12:39:14 DEBUG unit.mysql-k8s/0.juju-log ops 2.4.1 up and running.
unit-mysql-k8s-0: 12:39:14 DEBUG unit.mysql-k8s/0.juju-log load_ssl_context verify='/var/run/secrets/kubernetes.io/serviceaccount/ca.crt' cert=None trust_env=True http2=False
unit-mysql-k8s-0: 12:39:15 DEBUG unit.mysql-k8s/0.juju-log load_verify_locations cafile='/var/run/secrets/kubernetes.io/serviceaccount/ca.crt'
unit-mysql-k8s-0: 12:39:15 DEBUG unit.mysql-k8s/0.juju-log Invalid Loki alert rules folder at /var/lib/juju/agents/unit-mysql-k8s-0/charm/src/loki_alert_rules: directory does not exist
unit-mysql-k8s-0: 12:39:15 DEBUG unit.mysql-k8s/0.juju-log Emitting Juju event update_status.
unit-mysql-k8s-0: 12:39:15 INFO unit.mysql-k8s/0.juju-log Unit workload member-state is online with member-role primary
unit-mysql-k8s-0: 12:39:17 DEBUG unit.mysql-k8s/0.juju-log Getting cluster primary member's address from mysql-k8s-0.mysql-k8s-endpoints
unit-mysql-k8s-0: 12:39:19 DEBUG unit.mysql-k8s/0.juju-log Updating pod endpoint for mysql-k8s/0
[...skipped non-mysql-k8s-0 log lines...]
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log Labeling pod mysql-k8s-0 with label primary
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log connect_tcp.started host='10.108.0.1' port=443 local_address=None timeout=None socket_options=None
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log connect_tcp.complete return_value=<httpcore._backends.sync.SyncStream object at 0x7f67d0761f30>
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log start_tls.started ssl_context=<ssl.SSLContext object at 0x7f67d06fed40> server_hostname='10.108.0.1' timeout=None
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log start_tls.complete return_value=<httpcore._backends.sync.SyncStream object at 0x7f67d0762080>
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log send_request_headers.started request=<Request [b'GET']>
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log send_request_headers.complete
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log send_request_body.started request=<Request [b'GET']>
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log send_request_body.complete
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log receive_response_headers.started request=<Request [b'GET']>
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log receive_response_headers.complete return_value=(b'HTTP/1.1', 200, b'OK', [(b'Audit-Id', b'b4525a66-e88b-4136-8a34-9c8e8d0c86ad'), (b'Cache-Control', b'no-cache, private'), (b'Content-Type', b'application/json'), (b'X-Kubernetes-Pf-Flowschema-Uid', b'cadc6ce1-e5ad-4e8f-b9e7-62ca7fa65d1d'), (b'X-Kubernetes-Pf-Prioritylevel-Uid', b'74cb042e-3c8a-4780-9279-2a727532262b'), (b'Date', b'Mon, 10 Jul 2023 10:39:24 GMT'), (b'Transfer-Encoding', b'chunked')])
unit-mysql-k8s-0: 12:39:24 INFO unit.mysql-k8s/0.juju-log HTTP Request: GET https://10.108.0.1/api/v1/namespaces/mysql-k8s/pods/mysql-k8s-0 "HTTP/1.1 200 OK"
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log receive_response_body.started request=<Request [b'GET']>
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log receive_response_body.complete
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log response_closed.started
unit-mysql-k8s-0: 12:39:24 DEBUG unit.mysql-k8s/0.juju-log response_closed.complete
unit-mysql-k8s-0: 12:39:24 WARNING unit.mysql-k8s/0.juju-log No relation: certificates
unit-mysql-k8s-0: 12:39:25 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
[...skipped non-mysql-k8s-0 log lines...]
unit-mysql-k8s-0: 12:40:21 INFO juju.worker.caasunitterminationworker terminating due to SIGTERM <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
unit-mysql-k8s-0: 12:40:25 DEBUG unit.mysql-k8s/0.juju-log ops 2.4.1 up and running.
unit-mysql-k8s-0: 12:40:25 DEBUG unit.mysql-k8s/0.juju-log load_ssl_context verify='/var/run/secrets/kubernetes.io/serviceaccount/ca.crt' cert=None trust_env=True http2=False
unit-mysql-k8s-0: 12:40:25 DEBUG unit.mysql-k8s/0.juju-log load_verify_locations cafile='/var/run/secrets/kubernetes.io/serviceaccount/ca.crt'
unit-mysql-k8s-0: 12:40:25 DEBUG unit.mysql-k8s/0.juju-log Invalid Loki alert rules folder at /var/lib/juju/agents/unit-mysql-k8s-0/charm/src/loki_alert_rules: directory does not exist
unit-mysql-k8s-0: 12:40:25 DEBUG unit.mysql-k8s/0.juju-log Emitting Juju event stop.
unit-mysql-k8s-0: 12:40:26 INFO juju.worker.uniter.operation ran "stop" hook (via hook dispatching script: dispatch)
unit-mysql-k8s-0: 12:40:26 INFO juju.worker.uniter unit "mysql-k8s/0" shutting down: agent should be terminated
```

As you can see above, the `caasunitterminationworker` generates SIGTERM after about 1 minute of waiting after "update-status" ran.
It looks like we have some internal timer/timeout there.

The normal `update-status` case scenario (without performance test):
```
unit-mysql-k8s-0: 23:46:32 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
...
unit-mysql-k8s-0: 23:46:36 DEBUG unit.mysql-k8s/0.juju-log ops 2.4.1 up and running.
unit-mysql-k8s-0: 23:46:36 DEBUG unit.mysql-k8s/0.juju-log load_ssl_context verify='/var/run/secrets/kubernetes.io/serviceaccount/ca.crt' cert=None trust_env=True http2=False
unit-mysql-k8s-0: 23:46:36 DEBUG unit.mysql-k8s/0.juju-log load_verify_locations cafile='/var/run/secrets/kubernetes.io/serviceaccount/ca.crt'
unit-mysql-k8s-0: 23:46:36 DEBUG unit.mysql-k8s/0.juju-log Invalid Loki alert rules folder at /var/lib/juju/agents/unit-mysql-k8s-0/charm/src/loki_alert_rules: directory does not exist
...
unit-mysql-k8s-0: 23:46:36 DEBUG unit.mysql-k8s/0.juju-log Emitting Juju event update_status.
unit-mysql-k8s-0: 23:46:37 INFO unit.mysql-k8s/0.juju-log Alex Lutay: return doing nothing
```
As you can see, it took 4 seconds to do nothing (I have muted update_status charm block).
The line "ops 2.4.1 up and running" is missing in the logs above. Did ops fail to launch due to high unit/k8s/mysql load? 0_o

## Additional context
It looks very similar to https://github.com/canonical/prometheus-k8s-operator/issues/426 which came to nowhere...
Note we are using 2.9.44 with stop handling fixes, so https://bugs.launchpad.net/juju/+bug/1951415 is no longer a problem here!

The entire debug log with Juju 2,9 is [here](https://pastebin.canonical.com/p/kSBnyKTTqc/), see line 695 (SIGTERM time).
The entire debug log with Juju 3.2 is [here](https://pastebin.canonical.com/p/hgydCwptz4/), see line 493 (SIGTERM time).
Good ideas are welcome!

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

Mysqld should not use node total memory for parameters auto tuning.

## Solution

* Use k8s allocable memory as baseline for usable memory.
* consider `group_replication_message_cache_size` for calculating available pool size
* use smaller possible `group_replication_message_cache_size` when using `testing` profile

Extra: [The commit](https://github.com/canonical/mysql-k8s-operator/pull/256/commits/a1fb0332c24d9461f9ddead70e24bdb591b4ad30) address [DPE-2274](https://warthogs.atlassian.net/browse/DPE-2274) 

closes #254 

[DPE-2274]: https://warthogs.atlassian.net/browse/DPE-2274?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/mysql_k8s_helpers.py [hunk 1]
```diff
@@ -299,9 +301,13 @@ def create_custom_config_file(
 
         if innodb_buffer_pool_chunk_size:
             content.append(f"innodb_buffer_pool_chunk_size = {innodb_buffer_pool_chunk_size}")
-        content.append("")
+
+        # TODO: Enable after GR is enabled
```
Comment (taurus-forever) [inline src/mysql_k8s_helpers.py:305]: Remove TODO before merging?

**Review by taurus-forever — APPROVED** [review]
LGTM to merge this to continue testing performance using more stable MySQL mem usage.
One TODO should be removed/addressed before merging IMHO.

File: src/mysql_k8s_helpers.py (hunk 1 above)
Comment (paulomach) [reply src/mysql_k8s_helpers.py:305]: strangely enough I've removed and pushed, but the commit is not showing on gh

File: lib/charms/mysql/v0/mysql.py [hunk 2]
```diff
@@ -91,11 +91,14 @@ def wait_until_mysql_connection(self) -> None:
...
 # to 0 if you are raising the major API version
-LIBPATCH = 36
+LIBPATCH = 37
 
 UNIT_TEARDOWN_LOCKNAME = "unit-teardown"
 UNIT_ADD_LOCKNAME = "unit-add"
 
+BYTES_1GB = 1073741824  # 1 gibibyte
+BYTES_1Gb = 1000000000  # 1 gigabyte
+BYTES_1MB = 1048576     # 1 mebibyte
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py]: eb586a9 - good catch!

File: src/mysql_k8s_helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulomach) [reply src/mysql_k8s_helpers.py]: comment removed

File: src/mysql_k8s_helpers.py [hunk 3]
```diff
@@ -299,9 +301,13 @@ def create_custom_config_file(
...
+        if gr_message_cache_size:
+            content.append(f"loose-group_replication_message_cache_size = {gr_message_cache_size}")
 
         try:
+            content.append("")
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/mysql_k8s_helpers.py]: eb586a9

File: lib/charms/mysql/v0/mysql.py [hunk 4]
```diff
@@ -1558,7 +1565,7 @@ def get_max_connections(self) -> int:
         # Reference: based off xtradb-cluster-operator
         # https://github.com/percona/percona-xtradb-cluster-operator/blob/main/pkg/pxc/app/config/autotune.go#L61-L70
 
-        bytes_per_connection = 12582912  # 12 Megabytes
+        bytes_per_connection = 12 * BYTES_1MB  # 12 Megabytes
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py]: eb586a9

File: src/mysql_k8s_helpers.py (hunk 1 above)
Comment (paulomach) [reply src/mysql_k8s_helpers.py]: 33cbf57

**Pull request merged**